File: P4/Civitas/src/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author Salva
 */
public class TituloPropiedad {

    //atributos de instancia
    private float alquilerBase;
    private static final float factorInteresesHipoteca = 1.1f;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;

    //atributo de referencia
    private Jugador propietario;

    //metodos
    TituloPropiedad(String nombre, float alquBase, float factrevalor, float hipbase,
            float precioComp, float precioEdifi) {
        this.nombre = nombre;
        alquilerBase = alquBase;
        factorRevalorizacion = factrevalor;
        hipotecaBase = hipbase;
        precioCompra = precioComp;
        precioEdificar = precioEdifi;
        propietario = null;
        numCasas = numHoteles = 0;
        hipotecado = false;

    }

    @Override
    public String toString() {
        String salida = "Nombre: " + nombre
                + "\nAlquiler Base: " + alquilerBase
                + "\nFactor de Revalorización: " + factorRevalorizacion
                + "\nFactor Intereses Hipoteca: " + factorInteresesHipoteca
                + "\nHipoteca Base: " + hipotecaBase
                + "\nHipotecado: " + hipotecado
                + "\nNúmero de casas: " + numCasas
                + "\nNúmero de Hoteles: " + numHoteles
                + "\nPrecio de compra: " + precioCompra
                + "\nPrecio de Edificar: " + precioEdificar;
        return salida;

    }

    //cambiada visibilidad de private a paquete para que el jugador sepa de antemano la cantidad que debe pagar
    float getPrecioAlquiler() {
        //  PrecioAlquiler=AlquilerBase*(1+(NumCasas*0.5)+(NumHoteles*2.5))
        float precio = alquilerBase * (1f + (numCasas * 0.5f) + (numHoteles * 2.5f));
        if (hipotecado || propietarioEncarcelado()) {
            precio = 0;
        }
        return precio;
    }

    private float getImporteHipoteca() {
        //CantidadRecibida=HipotecaBase*(1+(numCasas*0.5)+(numHoteles*2.5))
        float importe = hipotecaBase * (1 + (numCasas * 0.5f) + (numHoteles * 2.5f));
        return importe;
    }

    float getImporteCancelarHipoteca() {
        float importe = getImporteHipoteca() * factorInteresesHipoteca;
        return importe;
    }

    void actualizaPropietarioPorConversion(Jugador jugador) {
        propietario = jugador;
    }

    boolean cancelarHipoteca(Jugador jugador) {
        boolean result = false;
        if (hipotecado) {
            if (esEsteElPropietario(jugador)) {
                jugador.paga(getImporteCancelarHipoteca());
                hipotecado = false;
                result = true;
            }
        }
        return result;
    }

    int cantidadCasasHoteles() {
        return getNumCasas() + getNumHoteles();
    }

    boolean comprar(Jugador jugador) {
        boolean result = false;
        if (!tienePropietario()) {
            propietario = jugador;
            result = true;
            propietario.paga(precioCompra);
        }
        return result;
    }

    boolean construirCasa(Jugador jugador) {
        boolean result = false;
        if (esEsteElPropietario(jugador)) {
            propietario.paga(precioEdificar);
            numCasas = numCasas + 1;
            result = true;
        }
        return result;
    }

    boolean construirHotel(Jugador jugador) {
        boolean result = false;
        if (esEsteElPropietario(jugador)) {
            propietario.paga(precioEdificar);
            numHoteles = numHoteles + 1;
            result = true;
        }
        return result;
    }

    boolean derruirCasas(int n, Jugador jugador) {
        if (esEsteElPropietario(jugador) && getNumCasas() >= n) {
            numCasas -= n;
            return true;
        }
        return false;
    }

    private boolean esEsteElPropietario(Jugador jugador) {
        return jugador == getPropietario();
    }

    public boolean getHipotecado() {
        return hipotecado;
    }

//    cambiada visibilidad de paquete a publica para vista textual
    public String getNombre() {
        return nombre;
    }

    int getNumCasas() {
        return numCasas;
    }

    int getNumHoteles() {
        return numHoteles;
    }

    float getPrecioCompra() {
        return precioCompra;
    }

    //modificamos visibilidad de paquete a public para que la partida sea más accesible. Se consulta al construir casas y hoteles
    public float getPrecioEdificar() {
        return precioEdificar;
    }

    private float getPrecioVenta() {
        float precio = getPrecioCompra() + getPrecioEdificar() * cantidadCasasHoteles() * factorRevalorizacion;
        return precio;
    }

    Jugador getPropietario() {
        return propietario;
    }

    boolean hipotecar(Jugador jugador) {
        boolean salida = false;
        if (!hipotecado && esEsteElPropietario(jugador)) {
            propietario.recibe(getImporteHipoteca());
            hipotecado = true;
            salida = true;
        }
        return salida;
    }

    private boolean propietarioEncarcelado() {
        return getPropietario().isEncarcelado();
    }

    boolean tienePropietario() {
        return propietario != null;
    }

    void tramitarAlquiler(Jugador jugador) {
        if (tienePropietario()) {
            if (!esEsteElPropietario(jugador)) {
                float precio = getPrecioAlquiler();
                jugador.paga(precio);
                propietario.recibe(precio);
            }
        }
    }

    boolean vender(Jugador jugador) {
        if (esEsteElPropietario(jugador) && !hipotecado) {
            jugador.recibe(getPrecioVenta());
            propietario = null;
            numCasas = numHoteles = 0;
            return true;
        } else {
            return false;
        }
    }

}

File: PruebaA2/civitasruby/pkg/civitasruby-0.0.1/lib/civitas/titulo_propiedad.rb
#encoding:utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class Titulo_propiedad
    #attr_reader :nombre, :precio_compra, :alquiler_base, :factor_revalorizacion,
    # :hipoteca_base, :precio_edificar, :num_hoteles, :num_casas
    attr_accessor :hipotecada, :propietario

    def initialize(nom, alquiler, factor, hipoteca, precioComp, precioEdif)
      @nombre = nom
      @precio_compra = precioComp
      @alquiler_base = alquiler
      @factor_revalorizacion = factor
      @hipoteca_base = hipoteca
      @precio_edificar = precioEdif
      @hipotecada = false
      @num_hoteles = 0
      @num_casas = 0
      @propietario = nil
    end

    @override
    def to_s()
      "\n-TituloPropiedad: \nNombre: #{@nombre} \nHipotecada: #{@hipotecada}
         \nPrecio compra: #{@precio_compra} \nAlquiler base: #{@alquiler_base}
         \nFactor revalorizacion: #{@factor_revalorizacion} \nHipoteca base: #{@hipoteca_base}
         \nPrecio edificar: #{@precio_edificar} \nNúmero de hoteles: #{@num_hoteles}
         \nNúmero de casas: #{@num_casas}\n\n"
    end

    def get_hipoteca_base
      @hipoteca_base
    end

    def get_num_casas
      @num_casas
    end

    def tiene_propietario
      @propietario != nil
    end

    def get_num_hoteles
      @num_hoteles
    end

    def get_alquiler_base
      @alquiler_base
    end

    def get_factor_revalorizacion
      @factor_revalorizacion
    end

    def get_precio_alquiler
      res = @alquiler_base * (1+(@num_casas*0.5) + (@num_hoteles*2.5))
      if @hipotecado || propietario_encarcelado
        res = 0
      end
      return res
    end

    def get_importe_hipoteca
      (@hipoteca_base*(1+(@num_casas*0.5) + (@num_hoteles*2.5)))
    end

    def get_importe_cancelar_hipoteca
      (get_importe_hipoteca * @@factor_intereses_hipoteca)
    end

    def actualiza_propietario_por_conversion(jugador)
      @propietario = jugador
    end

    def cancelar_hipoteca(jugador)
      result = false
      if(@hipotecado)
        if(es_este_el_propietaro(jugador))
          @propietario.paga(get_importe_cancelar_hipoteca)
          @hipotecado = false
          result =  true
        end
      end
      return result
    end


    def cantidad_casas_hoteles
      (get_num_casas + get_num_hoteles)
    end

    def comprar(jugador)
      result = false
      if (@propietario == nil)
        @propietario = jugador
        result = true
        @propietario.paga(@precio_compra)
      end
      return result
    end

    def construir_casa(jugador)
      result = false
      if(es_este_el_propietario(jugador))
        @propietario.paga(@precio_compra)
        @num_casas +=1
      end
    end

    def construir_hotel(jugador)
      result = false
      if (es_este_el_propietario(jugador))
        @propietario.paga(@precio_edificar)
        @num_hoteles = @num_hoteles + 1
        result = true
      end
      return result
    end

    def es_este_el_propietario(jugador)
      res = false
      if jugador == get_propietario
        res = true
      end
      return res
    end

    def get_hipotecado
      @hipotecado
    end

    def get_nombre
      @nombre
    end

    def get_num_casas
      @num_casas
    end

    def get_num_hoteles
      @num_hoteles
    end

    def get_precio_edificar
      @precio_edificar
    end

    def get_precio_compra
      @precio_compra
    end

    def get_precio_venta
      (get_precio_compra + get_precio_edificar*
        cantidad_casas_hoteles*@factor_revalorizacion)
    end

    def get_propietario
      @propietario
    end

    def hipotecar(jugador)
      salida = false
      if (!@hipotecado && es_este_el_propietario(jugador))
        @propietario.recibe(get_importe_hipoteca)
        @hipotecado = true
        salida = true
      end
      return salida
    end

    def tramitar_alquiler(jugador)
      if (@propietario != nil)
        if(es_este_el_propietario(jugador) == false)
          precio= get_precio_alquiler()
          jugador.paga_alquiler(precio)
          @propietario.recibe(precio)
        end
      end
    end

    def propietario_encarcelado
      res = false
      if @propietario.get_is_encarcelado == true
        res = true
      end
      return res
    end

    def cantidad_casas_hoteles
      (@num_casas + @num_hoteles)
    end

    def derruir_casas(n, jugador)
      res = false
      if es_este_el_propietario(jugador) && @num_casas >= n
        @num_casas = n
        res = true
      end
      return res
    end



    def vender(jugador)
      res = false
      if(!@hipotecado && es_este_el_propietario(jugador))
        jugador.recibe(get_precio_venta())
        derruir_casas(@num_casas,jugador)
        @num_hoteles = 0
        actualiza_propietario_por_conversion(nil)
        res = true
      end
      puts "\n**********\nvender de titulo: " + res.to_s + "\n**********\n"
      return res
    end

    private :es_este_el_propietario, :get_importe_hipoteca, :get_precio_venta #,  :get_precio_alquiler # este ultimo le cambiamos la visiblidad de private a public para mejorar la visibilidad del dato durante las partidas
  end
end
